### PR TITLE
Ote/fta

### PR DESCRIPTION
**Pull request by OrhanTekin**

### CODE REVIEW

File: extension/src/language-extension.ts [hunk 1]
```diff
@@ -170,7 +173,79 @@ export class StpaLspVscodeExtension extends SprottyLspEditVscodeExtension {
...
+                let id = "[";
+                for(const element of set){
+                    if(set.indexOf(element) === set.length -1){
+                        id += element.id;
+                    }else{
+                        id = id + element.id + ",";
+                    }
+                }
+                id += "]";
+                cutSetDropDownList.push({id: id, value: set});
```
Comment (Drakae) [inline extension/src/language-extension.ts:218]: Why do you need an ID in addition to the value?

File: extension/src/wview.ts [hunk 2]
```diff
@@ -71,6 +75,11 @@ export class StpaLspWebview extends SprottyLspWebview {
         const configOptions = vscode.workspace.getConfiguration('pasta');
         action.options.forEach(element => configOptions.update(element.id, element.value));
     }
+
+    protected selectCutSet(action: SelectCutSetAction):void{
+        this.dispatch(action);
+    }
+
```
Comment (Drakae) [inline extension/src/wview.ts:82]: not needed

File: extension/src-webview/options/cut-set-registry.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            for(const entry of action.cutSets){
+                dropDownOption.availableValues.push({displayName: entry.id , id: entry.id});
+            }
+
+            this._options.set('cut-sets', dropDownOption);
+            this.notifyListeners();
+        }else if(SelectCutSetAction.isThisAction(action)){
+            this.selectedCutSet = action.id;
+            this.ftaHightlighting = true;
+            this.highlightSelectedNodes(this.selectedCutSet);
```
Comment (Drakae) [inline extension/src-webview/options/cut-set-registry.ts:59]: not needed. Instead inject the registry in the "fta-views" file and ask for the value of the dropdown menu. Then you can highlight there the nodes and edges you want to highlight.

File: extension/src-webview/views.tsx [hunk 4]
```diff
@@ -199,8 +199,11 @@ export class STPAGraphView<IRenderingArgs> extends SGraphView<IRenderingArgs> {
         highlighting = allNodes.find(node => {
             return node instanceof STPANode && node.highlight
         }) !== undefined;
+        
+        setFTANodesAndEdges(allNodes, model.children as SEdge[]);
```
Comment (Drakae) [inline extension/src-webview/views.tsx:203]: not needed

**Review by Drakae — CHANGES_REQUESTED** [review]
Additionally to the comments, please merge the master into your branch.

File: extension/package.json [hunk 5]
```diff
@@ -85,254 +101,302 @@
...
+                "category": "Cut Sets"
             }
         ],
         "menus": {
             "commandPalette": [
                 {
-                    "command": "stpa.diagram.open",
-                    "when": "editorLangId == 'stpa'"
+                    "command": "pasta.diagram.open",
+                    "when": "editorLangId == 'stpa' || editorLangId == 'fta'"
```
Comment (Eddykasp) [inline extension/package.json:190]: see https://github.com/kieler/kieler-vscode/pull/75

File: extension/src-language-server/fta/analysis/fta-cutSet-calculator.ts [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    if (isComponent(startNode) || isCondition(startNode)) {
+        return [new Set<namedFtaElement>([startNode])];
+    }
+
+    const children = getChildrenOfNode(startNode);
+    if (children.length === 0 || !isGate(startNode)) {
+        return result;
+    }
+
+    if (isAND(startNode) || isInhibitGate(startNode)) {
```
Comment (Eddykasp) [inline extension/src-language-server/fta/analysis/fta-cutSet-calculator.ts:111]: this if(isA)... else if(isB) etc.. structure could be replaced by a switch case to be more readable I think

File: extension/src-language-server/fta/diagram/fta-layout-config.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+import { LayoutOptions } from "elkjs";
+import { DefaultLayoutConfigurator } from "sprotty-elk/lib/elk-layout";
+import { SGraph, SModelIndex, SNode } from "sprotty-protocol";
+
+export class FtaLayoutConfigurator extends DefaultLayoutConfigurator {
+    protected graphOptions(sgraph: SGraph, index: SModelIndex): LayoutOptions {
```
Comment (Eddykasp) [inline extension/src-language-server/fta/diagram/fta-layout-config.ts:23]: sgraph, index unused vars

File: extension/src-language-server/fta/diagram/fta-layout-config.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        return {
+            "org.eclipse.elk.direction": "DOWN",
+            "org.eclipse.elk.layered.nodePlacement.strategy": "NETWORK_SIMPLEX",
+        };
+    }
+
+    protected nodeOptions(snode: SNode, index: SModelIndex): LayoutOptions | undefined {
```
Comment (Eddykasp) [inline extension/src-language-server/fta/diagram/fta-layout-config.ts:30]: snode, index unused vars

File: extension/src-webview/fta/fta-views.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+import { FTAEdge, FTANode, FTNodeType } from './fta-model';
+
+@injectable()
+export class PolylineArrowEdgeViewFTA extends PolylineEdgeView {
+
+    protected renderLine(edge: FTAEdge, segments: Point[], context: RenderingContext): VNode {
+        const firstPoint = segments[0];
+        let path = `M ${firstPoint.x},${firstPoint.y}`;
+        for (let i = 1; i < segments.length; i++) {
+            const p = segments[i];
```
Comment (Eddykasp) [inline extension/src-webview/fta/fta-views.tsx:32]: what is a p?

File: extension/src-webview/options/render-options-registry.ts [hunk 10]
```diff
@@ -89,9 +89,8 @@ export class RenderOptionsRegistry extends Registry {
     handle(action: Action): void | Action | ICommand {
         if (SetRenderOptionAction.isThisAction(action)) {
             const option = this._renderOptions.get(action.id);
-
+
```
Comment (Eddykasp) [inline extension/src-webview/options/render-options-registry.ts:92]: remove whitespace

File: extension/src-webview/options/render-options-registry.ts [hunk 11]
```diff
@@ -110,7 +109,7 @@ export class RenderOptionsRegistry extends Registry {
                 option.currentValue = element.value;
             });
             this.notifyListeners();
-        }
+        }
```
Comment (Eddykasp) [inline extension/src-webview/options/render-options-registry.ts:112]: remove whitespace

File: extension/src-webview/views-rendering.tsx [hunk 12]
```diff
@@ -170,3 +175,93 @@ export function renderHexagon(node: SNode): VNode {
...
+ */
+export function renderAndGate(node: SNode): VNode {
+    const leftX = 0;
+    const midX = Math.max(node.size.width, 0) / 2.0;
+    const rightX = Math.max(node.size.width, 0);
+    const botY = Math.max(node.size.height, 0);
+    const midY = Math.max(node.size.height, 0) / 2.0;
+    const topY = 0;
+
+    const d = `M ${leftX}, ${midY} V ${botY} H ${rightX} V ${midY} C ${rightX}, ${midY} ${rightX}, ${topY} ${midX}, ${topY} ${leftX}, ${topY} ${leftX}, ${midY} ${leftX}, ${midY} Z`;
```
Comment (Eddykasp) [inline extension/src-webview/views-rendering.tsx:191]: what is a d?

File: extension/src-webview/views-rendering.tsx [hunk 13]
```diff
@@ -170,3 +175,93 @@ export function renderHexagon(node: SNode): VNode {
...
+    // 'M' + leftX + " " + midY + " V " + botY + " H " + rightX + " V " + midY + " C " + rightX + " " + midY + " " + rightX + " "
```
Comment (Eddykasp) [inline extension/src-webview/views-rendering.tsx:192]: is this comment helpful?

File: extension/src-webview/views-rendering.tsx [hunk 14]
```diff
@@ -170,3 +175,93 @@ export function renderHexagon(node: SNode): VNode {
...
+ */
+export function renderOrGate(node: SNode): VNode {
+    const leftX = 0;
+    const rightX = Math.max(node.size.width, 0);
+    const midX = rightX / 2.0;
+    const botY = Math.max(node.size.height, 0);
+    const nearBotY = botY - 5;
+    const midY = Math.max(node.size.height, 0) / 2;
+    const topY = 0;
+    const d = `M${leftX},${midY} V ${botY}` + `C ${leftX}, ${botY} ${leftX+10}, ${nearBotY} ${midX}, ${nearBotY} ${rightX-10}, ${nearBotY} ${rightX}, ${botY} ${rightX}, ${botY}`
```
Comment (Eddykasp) [inline extension/src-webview/views-rendering.tsx:213]: see above

File: extension/src-webview/views-rendering.tsx [hunk 15]
```diff
@@ -170,3 +175,93 @@ export function renderHexagon(node: SNode): VNode {
...
+ */
+export function renderKnGate(node: SNode, k: number, n: number): VNode {
+    const leftX = 0;
+    const rightX = Math.max(node.size.width, 0);
+    const midX = rightX / 2.0;
+    const botY = Math.max(node.size.height, 0);
+    const nearBotY = Math.max(node.size.height, 0) - (Math.max(node.size.height, 0) / 10.0);
+    const midY = Math.max(node.size.height, 0) / 2;
+    const topY = 0;
+    const d = `M${leftX},${midY} V ${botY}` + `C ${leftX}, ${botY} ${leftX+10}, ${nearBotY} ${midX}, ${nearBotY} ${rightX-10}, ${nearBotY} ${rightX}, ${botY} ${rightX}, ${botY}`
```
Comment (Eddykasp) [inline extension/src-webview/views-rendering.tsx:234]: see above, all these constructions are quite obscure, maybe a comment explaining what shape is being made would be helpful

File: extension/src-webview/views-rendering.tsx [hunk 16]
```diff
@@ -170,3 +175,93 @@ export function renderHexagon(node: SNode): VNode {
...
+export function renderInhibitGate(node: SNode): VNode {
+    const leftX = 0;
+    const midX = Math.max(node.size.width, 0) / 2.0;
+    const rightX = Math.max(node.size.width, 0);
+    const lowestY = Math.max(node.size.height, 0);
+    const lowY = Math.max(node.size.height, 0) - (Math.max(node.size.height, 0) / 4.0);
+    const highY = Math.max(node.size.height, 0) / 4.0;
+    const highestY = 0;
+
+    const d = 'M' + leftX + " " + lowY + " L " + leftX + " " + highY + " L " + midX + " " + highestY + " L " + rightX
```
Comment (Eddykasp) [inline extension/src-webview/views-rendering.tsx:261]: this is done differently than the previous ones

File: extension/src/language-extension.ts [hunk 17]
```diff
@@ -113,7 +113,10 @@ export class StpaLspVscodeExtension extends LspWebviewPanelManager {
         // send the changes to the language server
         const changes = changeEvent.contentChanges;
         const uri = changeEvent.document.uri.toString();
-        this.languageClient.sendNotification("editor/textChange", { changes: changes, uri: uri });
+        // TODO: ID enforcer for FTA
```
Comment (Eddykasp) [inline extension/src/language-extension.ts:116]: still todo?

**Review by Drakae — APPROVED** [review]

**Pull request merged**